dlt/sources/rest_api/typing.py

```diff
@@ -0,0 +1,279 @@
+    type: Optional[AuthType]  # noqa
+    token: str
+
+
+class ApiKeyAuthConfig(AuthTypeConfig, total=False):
+    """Uses provided `api_key` to create authorization data in the specified `location` (query, param, header, cookie) under specified `name`"""
+
+    name: Optional[str]
+    api_key: str
+    location: Optional[TApiKeyLocation]
+
+
+class HttpBasicAuthConfig(AuthTypeConfig, total=True):
+    """Uses HTTP basic authentication"""
+
+    username: str
+    password: str
+
+
+# TODO: add later
+# class OAuthJWTAuthConfig(AuthTypeConfig, total=True):
+
+
+AuthConfig = Union[
+    AuthConfigBase,
+    AuthType,
+    BearerTokenAuthConfig,
+    ApiKeyAuthConfig,
+    HttpBasicAuthConfig,
+    BearerTokenAuth,
+    APIKeyAuth,
+    HttpBasicAuth,
+]
+
+
+class ClientConfig(TypedDict, total=False):
+    base_url: str
+    headers: Optional[Dict[str, str]]
+    auth: Optional[AuthConfig]
+    paginator: Optional[PaginatorConfig]
+
+
+class IncrementalArgs(TypedDict, total=False):
+    cursor_path: str
+    initial_value: Optional[str]
+    last_value_func: Optional[LastValueFunc[str]]
+    primary_key: Optional[TTableHintTemplate[TColumnNames]]
+    end_value: Optional[str]
+    row_order: Optional[TSortOrder]
+    convert: Optional[Callable[..., Any]]
+
+
+class IncrementalConfig(IncrementalArgs, total=False):
+    start_param: str
+    end_param: Optional[str]
+
+
+ParamBindType = Literal["resolve", "incremental"]
+
+
+class ParamBindConfig(TypedDict):
+    type: ParamBindType  # noqa
+
+
+class ResolveParamConfig(ParamBindConfig):
+    resource: str
+    field: str
+
+
+class IncrementalParamConfig(ParamBindConfig, IncrementalArgs):
+    pass
+    # TODO: implement param type to bind incremental to
+    # param_type: Optional[Literal["start_param", "end_param"]]
+
+
+@dataclass
+class ResolvedParam:
+    param_name: str
+    resolve_config: ResolveParamConfig
+    field_path: jsonpath.TJsonPath = field(init=False)
+
+    def __post_init__(self) -> None:
+        self.field_path = jsonpath.compile_path(self.resolve_config["field"])
+
+
+class ResponseActionDict(TypedDict, total=False):
+    status_code: Optional[Union[int, str]]
+    content: Optional[str]
+    action: Optional[Union[str, Union[Callable[..., Any], List[Callable[..., Any]]]]]
+
+
+ResponseAction = Union[ResponseActionDict, Callable[..., Any]]
+
+
+class Endpoint(TypedDict, total=False):
+    path: Optional[str]
+    method: Optional[HTTPMethodBasic]
+    params: Optional[Dict[str, Union[ResolveParamConfig, IncrementalParamConfig, Any]]]
+    json: Optional[Dict[str, Any]]
+    paginator: Optional[PaginatorConfig]
+    data_selector: Optional[jsonpath.TJsonPath]
+    response_actions: Optional[List[ResponseAction]]
+    incremental: Optional[IncrementalConfig]
+
+
+class ProcessingSteps(TypedDict):
+    filter: Optional[Callable[[Any], bool]]  # noqa: A003
+    map: Optional[Callable[[Any], Any]]  # noqa: A003
+
+
+class ResourceBase(TypedDict, total=False):
```

Comment:
Merge with `TResourceHints`: extract common base to `TResourceHintsBase` and use it here

Reply:
Done by making the fields optional in the supertype.

Reply:
@rudolfix you'll have to verify that this is ok, I think it is